### DISCUSSION TITLE
кольцо воды

### DISCUSSION THREAD

**Discussion by Kor-Angar** [dosinabox/g2nr_unofficial_update#501] [General]
https://github.com/dosinabox/g2nr_unofficial_update/issues/325

https://github.com/dosinabox/g2nr_unofficial_update/issues/263

**Comment by Kor-Angar** [comment]
обучение у Гаана. если Гаану сначала поговорить про обучение, а потом показать кольцо, то обучение у Гаана будет стоить 100 золотых.

**Reply by Kor-Angar** [reply]
![20230212001044_1](https://user-images.githubusercontent.com/49514764/218286607-8fd4bfc1-65f0-4f18-acfe-928eb31f1aad.jpg)

**Reply by Kor-Angar** [reply]
![20230212001037_1](https://user-images.githubusercontent.com/49514764/218286616-0c499271-bf9f-4aec-8356-d198a00c3cea.jpg)

**Comment by Kor-Angar** [comment]
если у ITAR_Fake_RANGER сделать нулевую защиту, то в `B_MakeRangerReadyToLeaveMeeting` можно будет использовать
```D
        AI_EquipBestArmor(Ranger);
        Npc_RemoveInvItems(Ranger,ITAR_Fake_RANGER,1);
```

у меня сегодня корд опять оказался на ферме в доспехах Кольца Воды :-) так как Корд был одет в другие доспехи из моей сборки :-)

**Comment by Kor-Angar** [comment]
@dosinabox  в `B_EquipFakeRangerArmor` было бы хорошо добавить проверку на то какая броян одета
```D
func void B_EquipFakeRangerArmor(var C_Npc Ranger)
{
    var c_item Equiparmor; Equiparmor = Npc_GetEquippedArmor(ranger); 
    
    if (!Hlp_IsItem(Equiparmor, ITAR_Fake_RANGER)) // броня НЕОДЕТА
    {
        if(!Npc_GetInvItem(Ranger,ITAR_Fake_RANGER))
        {
            CreateInvItem(Ranger,ITAR_Fake_RANGER);
        };
        AI_EquipArmor(Ranger,item);
    };
};
```

**Reply by dosinabox** [reply]
```cpp
func void B_EquipFakeRangerArmor(var C_Npc ranger)
{
	if(!ArmorEquipped(ranger,ITAR_Fake_RANGER))
	{
		if(!Npc_HasItems(ranger,ITAR_Fake_RANGER))
		{
			CreateInvItems(ranger,ITAR_Fake_RANGER,1);
		};
		AI_EquipArmor(ranger,ITAR_Fake_RANGER);
	};
};
```

**Reply by Kor-Angar** [reply]
в таком случае проверка на Ларесе уже будет наверное не нужна в `B_MakeRangerReadyForMeeting`

**Reply by dosinabox** [reply]
тогда надо будет ещё проверять текущие доспехи в `B_MakeRangerReadyToLeaveMeeting`

**Reply by dosinabox** [reply]
Или нет. Короче, проверь пожалуйста:
```cpp
func void B_MakeRangerReadyForMeeting(var C_Npc ranger)
{
	if(Npc_IsDead(ranger))
	{
		return;
	};
	if(ArmorEquipped(ranger,ITAR_Fake_RANGER))
	{
		return;
	};
	if(!Npc_HasItems(ranger,ITAR_Fake_RANGER))
	{
		CreateInvItem(ranger,ITAR_Fake_RANGER);
	};
	AI_EquipArmor(ranger,ITAR_Fake_RANGER);
	ranger.npcType = NPCTYPE_FRIEND;
};

func void B_MakeRangerReadyToLeaveMeeting(var C_Npc ranger)
{
	if(Npc_IsDead(ranger))
	{
		return;
	};
	if(!ArmorEquipped(ranger,ITAR_Fake_RANGER))
	{
		return;
	};
	if(Hlp_GetInstanceID(ranger) == Hlp_GetInstanceID(Martin))
	{
		AI_EquipArmor(ranger,ITAR_MIL_L);
	}
	else if(Hlp_GetInstanceID(ranger) == Hlp_GetInstanceID(Lares))
	{
		AI_EquipArmor(ranger,ITAR_Vlk_L);
	}
	else if(Hlp_GetInstanceID(ranger) == Hlp_GetInstanceID(Cord))
	{
		AI_EquipArmor(ranger,ITAR_SLD_H);
	}
	else if(Hlp_GetInstanceID(ranger) == Hlp_GetInstanceID(Gaan))
	{
		AI_EquipArmor(ranger,ITAR_Bau_L);
	}
	else if(Hlp_GetInstanceID(ranger) == Hlp_GetInstanceID(Orlan))
	{
		AI_EquipArmor(ranger,ITAR_Bau_M);
	}
	else if(Hlp_GetInstanceID(ranger) == Hlp_GetInstanceID(Cavalorn))
	{
		if(MIS_Addon_Cavalorn_GetOrnamentFromPAL != FALSE)
		{
			AI_EquipArmor(ranger,ITAR_Bau_L);
		};
	};
};
```

**Reply by Kor-Angar** [reply]
я сегодня такое тестил активно:

```D
func void B_MakeRangerReadyForMeeting(var C_Npc Ranger)
{
	B_EquipFakeRangerArmor(Ranger);
    /*
	if(Hlp_GetInstanceID(Ranger) != Hlp_GetInstanceID(Lares))
	{
		B_EquipFakeRangerArmor(Ranger);
	}
	else if(LaresRangerArmorEquipped == FALSE)
	{
		B_EquipFakeRangerArmor(Ranger);
		LaresRangerArmorEquipped = TRUE;
	};
    */
	Ranger.npcType = NPCTYPE_FRIEND;
};

func void B_MakeRangerReadyToLeaveMeeting(var C_Npc Ranger)
{
	if (Hlp_GetInstanceID(Ranger) == Hlp_GetInstanceID(Cavalorn))
    && (MIS_Addon_Cavalorn_GetOrnamentFromPAL == true)
    {
        Npc_RemoveInvItem(Ranger,ITAR_RANGER_Addon); // лишняя броня у Каваларона которую не хочет удалять Дося
    };


    var c_item Equiparmor;  Equiparmor = Npc_GetEquippedArmor(ranger);
    if (Hlp_IsItem(Equiparmor, ITAR_Fake_RANGER))
    {
        // в версии Доси ITAR_Fake_RANGER имеет высокие показатели защиты
        Equiparmor.protection[PROT_EDGE]    = 0;
        Equiparmor.protection[PROT_BLUNT]   = 0;
        Equiparmor.protection[PROT_POINT]   = 0;
        Equiparmor.protection[PROT_FIRE]    = 0;
        Equiparmor.protection[PROT_MAGIC]   = 0;
        Equiparmor.value                    = 0;

        AI_EquipBestArmor(Ranger);                   // НПС одевают "родную" броню
        Npc_RemoveInvItem(Ranger,ITAR_Fake_RANGER);  // у НПС удаляется фейковая броня
    };

    /*
	if      (Hlp_GetInstanceID(Ranger) == Hlp_GetInstanceID(Martin))	{ AI_EquipArmor(Ranger,ITAR_MIL_L);	}
	else if (Hlp_GetInstanceID(Ranger) == Hlp_GetInstanceID(Cord))      { AI_EquipArmor(Ranger,ITAR_SLD_H);	}
	else if (Hlp_GetInstanceID(Ranger) == Hlp_GetInstanceID(Gaan))      { AI_EquipArmor(Ranger,ITAR_Bau_L);	}
	else if (Hlp_GetInstanceID(Ranger) == Hlp_GetInstanceID(Orlan))     { AI_EquipArmor(Ranger,ITAR_Bau_M);	}
    else if (Hlp_GetInstanceID(Ranger) == Hlp_GetInstanceID(Lares))	
    {	
        if(LaresRangerArmorEquipped == TRUE)		
        {
			AI_EquipArmor(Ranger,ITAR_Vlk_L);
			LaresRangerArmorEquipped = FALSE;
		};
	}
    else if(Hlp_GetInstanceID(Ranger) == Hlp_GetInstanceID(Cavalorn))
	{
		if(MIS_Addon_Cavalorn_GetOrnamentFromPAL != FALSE)
		{
			AI_EquipArmor(Ranger,ITAR_Bau_L);
		};
	};
    */
};
```

**Reply by dosinabox** [reply]
А что будет если целевой нпц мертв? Меняется ли доспех на трупе?

**Reply by Kor-Angar** [reply]
скорее всего будет меняться и на мёртвом. проверку на смерть нужно добавить в 
B_MakeRangerReadyForMeeting
B_MakeRangerReadyToLeaveMeeting

**Reply by dosinabox** [reply]
Код выше обновлен, добавлены проверки на смерть.
А проверка на надетую броню (моя `ArmorEquipped` или твоя) работает на нпц вне зоны видимости? Мне кажется что переменная `LaresRangerArmorEquipped` была нужна именно для такого случая.

**Reply by Kor-Angar** [reply]
По идеи должна работать и вне зоны видимости. Но в данном случае это не важно так как проверка стоит перед одеванием доспехом, а вот как раз на одевание доспеха  действуют ограничения.

переменную ты использовала что бы Ларес не раздевался при экипировки фейк доспеха, если уже в доспехе. я так понял что потом ты стала использовать разные доспехи: для походов оригинальные доспехи КВ, для собрание фейковые.

**Reply by dosinabox** [reply]
всегда фейковые

**Reply by Kor-Angar** [reply]
не могу найти где Ларес одевает доспехи при походи в чёрный лес

**Reply by Kor-Angar** [reply]
нашёл. у тебя Ларес переодевается через B_MakeRangerReadyToLeaveMeeting

**Reply by Kor-Angar** [reply]
я уже писал ранее, что для переодевание доспеха Ларесу / Каваларону / Грегу , лучше написать чёткую и понятную функцию, которая будет делать проверки,  удалять экипированную броню, создавать новую и одевать. наличие нескольких доспехов у НПС это плохо.

**Reply by Kor-Angar** [reply]
в любом случае Каваларону и Ларесу при нахождение за городом лучше сделать одинаковые доспехи. или фейковые или настоящие.

**Reply by dosinabox** [reply]
Кавалорна можно одевать в фейковые доспехи в его инстанции. Тогда потом не будет повторных.

**Reply by Kor-Angar** [reply]
или Лареса одевать в нормальные доспехи. тогда он не будет пытаться раздеться ;-)

**Reply by Kor-Angar** [reply]
если будешь делать какие то правки то напиши. что бы я мог свой код подогнать под твои правки.

**Reply by dosinabox** [reply]
> если будешь делать какие то правки то напиши. что бы я мог свой код подогнать под твои правки.

Пока только то что сверху, и только после тестирования.

**Comment by dosinabox** [comment]
аивар AIV_NPCIsRanger назначается, но никак не используется

**Reply by Kor-Angar** [reply]
не используется. но и смысла особого нет внедрять его.
